My Project/clsGenerateTranslations.vb

```diff
@@ -0,0 +1,580 @@
+    End Function
+
+    '''--------------------------------------------------------------------------------------------
+    ''' <summary>
+    ''' Returns a datatable with all form controls from the list of forms passed
+    ''' </summary>
+    ''' <param name="lstForms">The forms to get the controls for translations</param>
+    ''' <returns>A datatable with 3 columns; form_name, control_name, id_text. </returns>
+    '''--------------------------------------------------------------------------------------------
+    Public Shared Function GetControlsDatatable(lstForms As List(Of Form)) As DataTable
+        Dim clsDatatableControls As New DataTable
+        clsDatatableControls.Columns.Add("form_name", GetType(String))
+        clsDatatableControls.Columns.Add("control_name", GetType(String))
+        clsDatatableControls.Columns.Add("id_text", GetType(String))
+
+        For Each frm As Form In lstForms
+            Dim dctComponents As Dictionary(Of String, Component) = New Dictionary(Of String, Component)
+            clsWinFormsComponents.FillDctComponentsFromControl(frm, dctComponents)
+
+            For Each clsComponent In dctComponents
+                Dim idText As String
+
+                If TypeOf clsComponent.Value Is Control Then
+                    idText = GetActualTranslationText(DirectCast(clsComponent.Value, Control).Text)
+                ElseIf TypeOf clsComponent.Value Is ToolStripItem Then
+                    idText = GetActualTranslationText(DirectCast(clsComponent.Value, ToolStripItem).Text)
+                Else
+                    Throw New Exception("Developer Error: Translation dictionary entry (" & frm.Name & "," & clsComponent.Key & ") contained unexpected value type.")
+                    Exit For
+                End If
+                'add row of form_name, control_name, id_text
+                clsDatatableControls.Rows.Add(frm.Name, clsComponent.Key, idText)
+            Next
+
+            'Special case for radio buttons in panels: 
+            '  Before the dialog is shown, each radio button is a direct child of the dialog 
+            '  (e.g. 'dlg_Augment_rdoNewDataframe'). After the dialog is shown, the raio button becomes 
+            '  a direct child of its parent panel.
+            '  Therefore, we need to show the dialog before we traverse the dialog's control hierarchy.
+            '  Unfortunately showing the dialog means that it has to be manually closed. So we only 
+            '  show the dialog for this special case to save the developer from having to manually 
+            '  close too many dialogs.
+            '  TODO: launch each dialog in a new thread to avoid need for manual close?
+            'If strTemp.ToLower().Contains("pnl") AndAlso strTemp.ToLower().Contains("rdo") Then
+            '    'frmTemp.ShowDialog()
+            '    frmTemp.Show()
+            '    strTemp = GetControls(frmTemp)
+            '    frmTemp.Close()
+            'End If
```

Comment:
I assume that this is commented out because it's an R-Instat specific problem that does not occur in Climsoft, is that correct?
If so, should we delete this text rather than have it commented out?

Reply:
@lloyddewit let's leave this for reference. I'm trying to find a way of seeing how we can get the texts of controls that have their contents set during the runtime. In climsoft, this is for controls like datagridview, list boxes headers and contents coming from database. In probably my next PR I will remove this comment.

Reply:
Understood, I set this conversation to resolved